File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    let rows = matrix.length
    let cols = matrix[0].length

    let s = 0
    let e = rows * cols -1

    while(s<=e){
        let mid = Math.floor(s + (e-s)/2)

        let r = Math.floor(mid / cols)
        let c = mid % cols

        if(matrix[r][c] === target) return true
        else if(matrix[r][c] > target) e = mid-1
        else s = mid+1
    }

    return false
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {\
        int rows = matrix.size();
        int cols = matrix[0].size();
        int i = 0;
        int j = rows * cols - 1;

        while(i<=j){
            int mid = i + (j-i)/2;

            int r = mid / cols;
            int c = mid % cols;

            if(matrix[r][c] == target) return true;
            else if(matrix[r][c] > target) j = mid-1;
            else i = mid+1;
        }

        return false;
    }
};